### PR TITLE
added passing tests by a student(with time control) and checking result of test. 

### PR DESCRIPTION
**Pull request by Yurii-Romaniv**
added passing tests by a student(with time control) and checking result of test.  
when the student starts the test, his `TestSessionInfo` is updated (the id of the test the student is taking and the time by which he should send the results are recorded there. `INSURANCE_TIME` is added before this time so that the test is not closed prematurely due to network delays).

### CODE REVIEW

File: src/main/java/com/example/nlearn/services/TestPassingService.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    @Autowired
+    private MarkService markService;
+
+
+    public void raiseErrorIfTestingNotAllowed(Test test, int testId, User user, TestSessionInfo sessionInfo, LocalDateTime currentTime, LocalDateTime endTime) {
+        boolean isTestForThisStudent = test.getGroup().getId() == user.getGroup().getId();
+        boolean areRetriesOver = markService.countMarksByUserAndTest(user, test) >= test.getNumberOfRetries();
+        if ((sessionInfo.isActive() && sessionInfo.getCurrentTestId() != testId) || !isTestForThisStudent || endTime.isBefore(currentTime) || areRetriesOver) {
+            throw new ResourceAccessException("");
```
Comment (justme1135) [inline src/main/java/com/example/nlearn/services/TestPassingService.java]: Create custom exception for that

File: src/main/java/com/example/nlearn/services/TestPassingService.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        }
+    }
+
+
+    @Transactional(propagation = Propagation.NOT_SUPPORTED)
+    public FullTest startTest(Integer testId, User user) {
+        Test test = testRepository.getTestById(testId);
+        TestSessionInfo sessionInfo = user.getTestSessionInfo();
+        LocalDateTime currentTime = java.time.LocalDateTime.now();
```
Comment (justme1135) [inline src/main/java/com/example/nlearn/services/TestPassingService.java]: why do you need to pass the package here?

File: src/main/java/com/example/nlearn/services/QuestionService.java [hunk 3]
```diff
@@ -35,5 +35,11 @@ public Question getById(Integer id) {
     public List<Question> findByTestId(Integer id) {
         return questionRepository.findByTestIdOrderById(id);
     }
+
+    public List<Question> findByTestIdWithoutAnswers(Integer id) {
+        List<Question> questions = questionRepository.getByTestIdOrderById(id);
```
Comment (justme1135) [inline src/main/java/com/example/nlearn/services/QuestionService.java:40]: In order not to have questions from DB modified you may use smth like that:
`ArrayList<Question> questions = new ArrayList(questionRepository.getByTestIdOrderById(id))`
This way you might be able to remove `@Transactional(propagation = Propagation.NOT_SUPPORTED)`

File: src/main/java/com/example/nlearn/services/TestPassingService.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+        List<Question> questions = questionService.findByTestId(testId);
+
+        double testValuation = StreamUtils.zip(answeredQuestions.stream(), questions.stream(), (answered, real) -> {
+            if (answered.getCorrectIndexes().size() > real.getCorrectIndexes().size()) {
+                return 0.0;
+            }
+
+            return answered.getCorrectIndexes().stream().mapToInt(integer -> {
+                if (real.getCorrectIndexes().contains(integer)) {
```
Comment (justme1135) [inline src/main/java/com/example/nlearn/services/TestPassingService.java]: 1) You may consider using a ternary operator here
2) You might use mapToDouble and not cast to double in line 95

File: src/main/java/com/example/nlearn/controllers/StudentsHomeController.java [hunk 5]
```diff
@@ -18,5 +25,22 @@ public class StudentsHomeController {
...
+    @Autowired
+    private TestPassingService testPassingService;
+
+
+    @GetMapping("/load-tests")
+    public StudentsContent home(@AuthenticationPrincipal CustomOAuth2User user) {
+        return testService.getContentForStudent(user.getDbUser());
+    }
+
+    @GetMapping("/start-test/{id}")
```
Comment (justme1135) [inline src/main/java/com/example/nlearn/controllers/StudentsHomeController.java]: `/{id}/start`

File: src/main/java/com/example/nlearn/controllers/StudentsHomeController.java [hunk 6]
```diff
@@ -18,5 +25,22 @@ public class StudentsHomeController {
...
+    public FullTest testing(@PathVariable Integer id, @AuthenticationPrincipal CustomOAuth2User user) {
+        return testPassingService.startTest(id, user.getDbUser());
+    }
+
+    @PostMapping("/end-test/{id}")
```
Comment (justme1135) [inline src/main/java/com/example/nlearn/controllers/StudentsHomeController.java]: `/{id}/end`

File: src/main/java/com/example/nlearn/services/TestPassingService.java (hunk 1 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/services/TestPassingService.java]: done

File: src/main/java/com/example/nlearn/services/TestPassingService.java (hunk 2 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/services/TestPassingService.java]: fixed

File: src/main/java/com/example/nlearn/services/QuestionService.java (hunk 3 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/services/QuestionService.java:40]: this solution doesn't work

File: src/main/java/com/example/nlearn/services/TestPassingService.java (hunk 4 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/services/TestPassingService.java]: 1. done 
2.  done

File: src/main/java/com/example/nlearn/controllers/StudentsHomeController.java (hunk 5 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/controllers/StudentsHomeController.java]: done

File: src/main/java/com/example/nlearn/controllers/StudentsHomeController.java (hunk 6 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/controllers/StudentsHomeController.java]: done

File: src/main/java/com/example/nlearn/services/QuestionService.java (hunk 3 above)
Comment (justme1135) [reply src/main/java/com/example/nlearn/services/QuestionService.java:40]: It's really strange. Try something of the https://www.baeldung.com/java-copy-list-to-another to be 100% sure

**Review by justme1135 — COMMENTED** [review]
I recommend you to debug this issue with questions being mutated because it is a strange behavior. Apart from that you are good to go

**Comment by Yurii-Romaniv** [comment]
> I recommend you to debug this issue with questions being mutated because it is a strange behavior. Apart from that you are good to go

I tried. copying the array / moving this code in other methods did not help

**Review by justme1135 — APPROVED** [review]

**Pull request merged**